### PR TITLE
fix(langchain): safely check if instance is openai llm/chat

### LINKED ISSUE #8889: [DDTrace==2.8.0] AttributeError: module 'langchain_community' has no attribute 'chat_models'
### Summary of problem
We are using LangChain with the ChatOpenAI module. When running ChatOpenAI, we get a datadog error with 
```
AttributeError: module 'langchain_community' has no attribute 'chat_models'
```
This comes from 
```
ddtrace/contrib/langchain/patch.py:349
```
It seems like langchain-community <= 0.0.10 puts the chat_models in langchain-community, while > 0.0.10 instructs you to pip install `langchain-openai` and import from langchain_openai.
But datadog's codebase says that it should look at langchain_community for that file path.

In ddtrace, this line is the offender:
```
if isinstance(instance, BASE_LANGCHAIN_MODULE.chat_models.ChatOpenAI) or (
            langchain_openai and isinstance(instance, langchain_openai.ChatOpenAI)
        ):
```
The problem is that when checking the isinstance, python tries to complete the import. I believe the easiest solution is to just switch the two boolean checks.

```
if (
            langchain_openai and isinstance(instance, langchain_openai.ChatOpenAI)
        ) or isinstance(instance, BASE_LANGCHAIN_MODULE.chat_models.ChatOpenAI):
```
This wont fix the underlying issue of trying to grab an import when it does not exist, but it will not break it for the rest of us :)

Here is a proposed solution, I will also make a Pull Request if possible:
```
import importlib
def safe_isinstance(instance, base_module, module_suffix):
    """
    Checks if the module exists safely without throwing an error
    """
    base_module_name = getattr(base_module, "__name__")
    try:
        module = importlib.import_module(f"{base_module_name}.{module_suffix}")
    except (ModuleNotFoundError, ImportError):
        return False

    return isinstance(instance, module)

if safe_isinstance(instance, BASE_LANGCHAIN_MODULE, "chat_models.ChatOpenAI") or (
    langchain_openai and safe_isinstance(instance, langchain_openai, "ChatOpenAI")
):
```
And that should safely check if the module exists first and then check if the instance is an instance of that module so we fail gracefully

### Which version of dd-trace-py are you using?
`ddtrace==2.8.0`

### Which version of pip are you using?
`pip 24.0 from /usr/local/lib/python3.11/site-packages/pip (python 3.11)`

### Which libraries and their versions are you using?

<details>
  <summary>`pip freeze`</summary>
  ```
ddtrace==2.8.0
langchain==0.1.14
langchain-community==0.0.31
langchain-core==0.1.40
langchain-openai==0.1.1
langchain-pinecone==0.0.3
langchain-text-splitters==0.0.1
langdetect==1.0.9
langsmith==0.1.40
  ```
</details>

### How can we reproduce your problem?
We are using a docker container for python3.11, but you can use any linux distro (or local)
```
pip install ddtrace==2.8.0 langchain langchain-openai langchain-community python-dotenv
```
This requires a ddtrace api key and openai api key in the env vars
For reproduction, im using a simple 
```
ddtrace-run ipython
```
And then run 
```
from langchain_openai import ChatOpenAI
from langchain.prompts import ChatPromptTemplate
from langchain_core.output_parsers import StrOutputParser
from langchain_core.runnables import RunnablePassthrough
from dotenv import load_dotenv
load_dotenv()
model = ChatOpenAI(model="gpt-4-turbo-preview")
chat_prompt = ChatPromptTemplate.from_messages(["What is 1 + 1?"])
retrieval_chain = (
    {"test": RunnablePassthrough()}
    | chat_prompt
    | model
    | StrOutputParser()
)
retrieval_chain.invoke("test")
```

### What is the result that you get?
And that raises
```
Error submitting packet: [Errno 111] Connection refused, dropping the packet and closing the socket
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[1], line 15
      8 chat_prompt = ChatPromptTemplate.from_messages(["What is 1 + 1?"])
      9 retrieval_chain = (
     10     {"test": RunnablePassthrough()}
     11     | chat_prompt
     12     | model
     13     | StrOutputParser()
     14 )
---> 15 retrieval_chain.invoke("test")

File /usr/local/lib/python3.11/site-packages/langchain_core/runnables/base.py:2499, in RunnableSequence.invoke(self, input, config)
   2497 try:
   2498     for i, step in enumerate(self.steps):
-> 2499         input = step.invoke(
   2500             input,
   2501             # mark each step as a child run
   2502             patch_config(
   2503                 config, callbacks=run_manager.get_child(f"seq:step:{i+1}")
   2504             ),
   2505         )
   2506 # finish the root run
   2507 except BaseException as e:

File /usr/local/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py:154, in BaseChatModel.invoke(self, input, config, stop, **kwargs)
    143 def invoke(
    144     self,
    145     input: LanguageModelInput,
   (...)
    149     **kwargs: Any,
    150 ) -> BaseMessage:
    151     config = ensure_config(config)
    152     return cast(
    153         ChatGeneration,
--> 154         self.generate_prompt(
    155             [self._convert_input(input)],
    156             stop=stop,
    157             callbacks=config.get("callbacks"),
    158             tags=config.get("tags"),
    159             metadata=config.get("metadata"),
    160             run_name=config.get("run_name"),
    161             **kwargs,
    162         ).generations[0][0],
    163     ).message

File /usr/local/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py:554, in BaseChatModel.generate_prompt(self, prompts, stop, callbacks, **kwargs)
    546 def generate_prompt(
    547     self,
    548     prompts: List[PromptValue],
   (...)
    551     **kwargs: Any,
    552 ) -> LLMResult:
    553     prompt_messages = [p.to_messages() for p in prompts]
--> 554     return self.generate(prompt_messages, stop=stop, callbacks=callbacks, **kwargs)

File /usr/local/lib/python3.11/site-packages/ddtrace/contrib/trace_utils.py:334, in with_traced_module.<locals>.with_mod.<locals>.wrapper(wrapped, instance, args, kwargs)
    332     log.debug("Pin not found for traced method %r", wrapped)
    333     return wrapped(*args, **kwargs)
--> 334 return func(mod, pin, wrapped, instance, args, kwargs)

File /usr/local/lib/python3.11/site-packages/ddtrace/contrib/langchain/patch.py:349, in traced_chat_model_generate(langchain, pin, func, instance, args, kwargs)
    346         span.set_tag_str("langchain.request.%s.parameters.%s" % (llm_provider, param), str(val))
    348 chat_completions = func(*args, **kwargs)
--> 349 if isinstance(instance, BASE_LANGCHAIN_MODULE.chat_models.ChatOpenAI) or (
    350     langchain_openai and isinstance(instance, langchain_openai.ChatOpenAI)
    351 ):
    352     _tag_openai_token_usage(span, chat_completions.llm_output)
    353     integration.record_usage(span, chat_completions.llm_output)

AttributeError: module 'langchain_community' has no attribute 'chat_models'
```
### What is the result that you expected?
Not to raise an exception

### Stack Trace
```
Traceback (most recent call last):
fastapi-1        |   File "/app/qclippy/fastapi/main.py", line 100, in ask_question
fastapi-1        |     chain_answer = gpt_qa_chain.invoke(
fastapi-1        |                    ^^^^^^^^^^^^^^^^^^^^
fastapi-1        |   File "/usr/local/lib/python3.11/site-packages/langchain_core/runnables/base.py", line 2499, in invoke
fastapi-1        |     input = step.invoke(
fastapi-1        |             ^^^^^^^^^^^^
fastapi-1        |   File "/usr/local/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 154, in invoke
fastapi-1        |     self.generate_prompt(
fastapi-1        |   File "/usr/local/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 554, in generate_prompt
fastapi-1        |     return self.generate(prompt_messages, stop=stop, callbacks=callbacks, **kwargs)
fastapi-1        |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
fastapi-1        |   File "/usr/local/lib/python3.11/site-packages/ddtrace/contrib/trace_utils.py", line 334, in wrapper
fastapi-1        |     return func(mod, pin, wrapped, instance, args, kwargs)
fastapi-1        |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
fastapi-1        |   File "/usr/local/lib/python3.11/site-packages/ddtrace/contrib/langchain/patch.py", line 349, in traced_chat_model_generate
fastapi-1        |     if isinstance(instance, BASE_LANGCHAIN_MODULE.chat_models.ChatOpenAI) or (
fastapi-1        |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
fastapi-1        | AttributeError: module 'langchain_community' has no attribute 'chat_models'
```

### PR DESCRIPTION
**Pull request by Yun-Kim**
Fixes #8889.

This PR adds a safe check for if a traced LLM/Chat model instance is an OpenAI instance by adding try/catch and reversing the order of the type checks. Previously we were checking directly against the installed langchain package, i.e.
```python
if isinstance(instance, BASE_LANGCHAIN_MODULE.chat_models.ChatOpenAI) or (
            langchain_openai and isinstance(instance, langchain_openai.ChatOpenAI)
        ):
# BASE_LANGCHAIN_MODULE can be either `langchain` or `langchain_community`
```
But `langchain_community` does not allow automatically accessing its submodules, i.e. `langchain_community.chat_models.ChatOpenAI` will result in an error unless `from langchain_community import chat_models` is performed already.

With this fix, there are three scenarios for `langchain_community/langchain_openai` users:
1. They use `langchain_openai` --> we perform the type check using `langchain_openai` first which will always be available, and will never hit the `BASE_LANGCHAIN_MODULE` type check.
2. They use `langchain_community` --> since users are using `langchain_community.chat_models` they must have already imported this in their code and it should not result in any errors. Regardless, we will safely try type checking against the submodule.
3. They use `langchain` --> `langchain` allows automatically accessing submodules without directly importing, so this should also not result in any errors.

## Checklist

- [x] Change(s) are motivated and described in the PR description
- [x] Testing strategy is described if automated tests are not included in the PR
- [x] Risks are described (performance impact, potential for breakage, maintainability)
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [x] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Description motivates each change
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] Release note makes sense to a user of the library
- [x] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

**Review by ZStriker19 — COMMENTED** [review]
Should we have a regression test?

**Comment by Yun-Kim** [comment]
> Should we have a regression test?

Honestly this is a really weird edge case, because we're dealing with package-level imports. We have test cases covering `langchain<=0.0.354` and `langchain>=0.1.0`, but importing `chat_models/llms` from `langchain_community` is actually already deprecated and slated to be removed in `langchain>=0.2.0`. I'm not sure how valuable it would be to add another venv to test (increase CI time by 2x) for the sake of testing an edge case in langchain's transition to a stable version.

File: ddtrace/contrib/langchain/patch.py [hunk 1]
```diff
@@ -145,6 +145,30 @@ def _tag_openai_token_usage(
         _tag_openai_token_usage(span._parent, llm_output, propagated_cost=propagated_cost + total_cost, propagate=True)
 
 
+def _is_openai_llm_instance(instance):
+    """Safely check if a traced instance is an OpenAI LLM.
+    langchain_community does not automatically import submodules which may result in AttributeErrors.
```
Comment (davidallada) [inline ddtrace/contrib/langchain/patch.py:150]: My understanding is that langchain_community deprecated the chat_models submodule, not necessarily that they do not automatically import submodules.

File: ddtrace/contrib/langchain/patch.py [hunk 2]
```diff
@@ -145,6 +145,30 @@ def _tag_openai_token_usage(
...
+        if langchain_openai:
+            return isinstance(instance, langchain_openai.OpenAI)
+        return isinstance(instance, BASE_LANGCHAIN_MODULE.llms.OpenAI)
+    except (AttributeError, ModuleNotFoundError, ImportError):
+        return False
+
+
+def _is_openai_chat_instance(instance):
+    """Safely check if a traced instance is an OpenAI Chat Model.
+    langchain_community does not automatically import submodules which may result in AttributeErrors.
```
Comment (davidallada) [inline ddtrace/contrib/langchain/patch.py:162]: same comment here
```
My understanding is that langchain_community deprecated the chat_models submodule, not necessarily that they do not automatically import submodules.
```

**Review by davidallada — APPROVED** [review]
Thanks for the quick turnaround on this patch to our issue.
The logic looks good here, I just made a very small comment about the two comments for the _is functions

File: ddtrace/contrib/langchain/patch.py (hunk 1 above)
Comment (Yun-Kim) [reply ddtrace/contrib/langchain/patch.py:150]: From what I understand, `langchain_community` hasn't deprecated the entire `chat_models` submodule but is rather deprecating individual classes (ex: `ChatOpenAI`) as they work on setting up partner libraries (ex: `langchain_openai`). Also, `langchain_community.__init__.py` does not automatically import its submodules, i.e.

```python
>>> import langchain_community
>>> langchain_community.llms  # try referencing directly, results in error
AttributeError: module 'langchain_community' has no attribute 'chat_models'
>>> from langchain_community import llms  # works
```
